File: variant-a/csharp-service/Services/OrderService.cs
using AutoMapper;
using FlashSale.Api.Data;
using FlashSale.Api.DTOs;
using FlashSale.Api.Models;
using FlashSaleAPI.Services;
using Microsoft.EntityFrameworkCore;

namespace FlashSale.Api.Services;

public class OrderService : IOrderService
{
    private readonly FlashSaleDbContext _context;
    private readonly IMapper _mapper;
    private readonly ILogger<OrderService> _logger;
    private readonly CSharpSnowflakeGenerator _idGenerator;
    private readonly RedisCacheService _redisCache;
    private readonly CampaignMemoryAllocator _campaignAllocator;

    public OrderService(FlashSaleDbContext context, IMapper mapper, ILogger<OrderService> logger, CSharpSnowflakeGenerator idGenerator, RedisCacheService redisCache, CampaignMemoryAllocator campaignAllocator)
    {
        _context = context;
        _mapper = mapper;
        _logger = logger;
        _idGenerator = idGenerator;
        _redisCache = redisCache;
        _campaignAllocator = campaignAllocator;
    }

    public async Task<IEnumerable<OrderResponseDto>> GetAllAsync(int skip = 0, int take = 100, string? customerEmail = null)
    {
        var query = _context.Orders
            .Include(o => o.LineItems)
            .Include(o => o.FlashSaleCampaign)
            .AsQueryable();

        if (!string.IsNullOrEmpty(customerEmail))
        {
            query = query.Where(o => o.CustomerEmail.Contains(customerEmail));
        }

        var orders = await query
            .OrderByDescending(o => o.CreatedAt)
            .Skip(skip)
            .Take(take)
            .ToListAsync();

        return _mapper.Map<IEnumerable<OrderResponseDto>>(orders);
    }

    public async Task<OrderResponseDto?> GetByIdAsync(Guid id)
    {
        var order = await _context.Orders
            .Include(o => o.LineItems)
            .Include(o => o.FlashSaleCampaign)
            .FirstOrDefaultAsync(o => o.Id == id);

        return order == null ? null : _mapper.Map<OrderResponseDto>(order);
    }

    /// <summary>
    /// Create order with intelligent routing:
    /// - If SKU is in active flash sale campaign → Use Variant A (Allocation-based adaptive inventory)
    /// - Otherwise → Use Variant Y (database transaction)
    /// Frontend sees same API, backend handles routing transparently.
    /// </summary>
    public async Task<OrderResponseDto> CreateAsync(OrderCreateDto dto)
    {
        var orderNumber = $"ORD-{_idGenerator.Generate()}";

        // Step 1: Check if ANY SKU is in active flash sale campaign
        var skuIds = dto.LineItems.Select(item => item.SkuId).ToList();

        Guid? flashSaleId = null;
        bool useVariantA = false;

        foreach (var skuId in skuIds)
        {
            var meta = await _redisCache.GetSkuMetaAsync(skuId);
            if (meta != null && meta.ContainsKey("flash_sale_id") && meta.GetValueOrDefault("status") == "active")
            {
                flashSaleId = Guid.Parse(meta["flash_sale_id"]);
                useVariantA = true;
                _logger.LogInformation("SKU {SkuId} is in active flash sale {FlashSaleId}, using Variant A", skuId, flashSaleId);
                break;
            }
        }

        if (useVariantA && flashSaleId.HasValue)
        {
            // VARIANT A: Allocation-based Adaptive Inventory (Flash Sale Path)
            return await CreateOrderVariantAAsync(dto, orderNumber, flashSaleId.Value);
        }
        else
        {
            // VARIANT Y: Database Transaction (Regular Order Path)
            return await CreateOrderVariantYAsync(dto, orderNumber);
        }
    }

    /// <summary>
    /// Variant Y: Traditional database transaction path (4-7 queries per order).
    /// Used for regular orders when SKU is NOT in active flash sale.
    /// </summary>
    private async Task<OrderResponseDto> CreateOrderVariantYAsync(OrderCreateDto dto, string orderNumber)
    {
        using var transaction = await _context.Database.BeginTransactionAsync();

        try
        {
            _logger.LogInformation("Using Variant Y (database) for order {OrderNumber}", orderNumber);

            // Create order
            var order = new Order
            {
                OrderNumber = orderNumber,
                CustomerEmail = dto.CustomerEmail,
                CustomerName = dto.CustomerName,
                TaxAmount = dto.TaxAmount,
                ShippingAmount = dto.ShippingAmount,
                Currency = dto.Currency,
                Notes = dto.Notes,
                FlashSaleCampaignId = dto.FlashSaleCampaignId
            };

            _context.Orders.Add(order);
            await _context.SaveChangesAsync(); // Get the order ID

            // Add line items and calculate totals
            decimal subtotal = 0;
            foreach (var itemDto in dto.LineItems)
            {
                // Get SKU information
                var sku = await _context.Skus
                    .Include(s => s.Inventory)
                    .Include(s => s.Spu)
                    .FirstOrDefaultAsync(s => s.Id == itemDto.SkuId);

                if (sku == null)
                {
                    throw new InvalidOperationException($"SKU {itemDto.SkuId} not found");
                }

                // Check inventory
                if (sku.TrackInventory && sku.Inventory != null)
                {
                    if (!sku.Inventory.CanFulfillQuantity(itemDto.Quantity))
                    {
                        throw new InvalidOperationException($"Insufficient inventory for SKU {sku.SkuCode}");
                    }

                    // Reserve inventory
                    sku.Inventory.ReserveQuantity(itemDto.Quantity);
                }

                // Check for active flash sale campaign and apply flash price
                Models.FlashSale? activeCampaign = null;
                if (sku.SpuId != null)
                {
                    var now = DateTime.UtcNow;
                    activeCampaign = await _context.FlashSaleCampaigns
                        .FirstOrDefaultAsync(c =>
                            c.SpuId == sku.SpuId &&
                            c.IsActive &&
                            c.Status == "active" &&
                            c.StartTime <= now &&
                            c.EndTime >= now);
                }

                // Create line item - use flash price if campaign is active
                var unitPrice = activeCampaign != null
                    ? (itemDto.UnitPrice ?? activeCampaign.FlashPrice)
                    : (itemDto.UnitPrice ?? sku.Price);
                var totalPrice = unitPrice * itemDto.Quantity;

                var lineItem = new OrderLineItem
                {
                    OrderId = order.Id,
                    SkuId = sku.Id,
                    Quantity = itemDto.Quantity,
                    UnitPrice = unitPrice,
                    TotalPrice = totalPrice,
                    ProductName = sku.Spu?.Name ?? sku.Name ?? "Product",
                    SkuCode = sku.SkuCode
                };

                _context.OrderLineItems.Add(lineItem);
                subtotal += totalPrice;
            }

            // Update order totals
            order.Subtotal = subtotal;
            order.TotalAmount = subtotal + order.TaxAmount + order.ShippingAmount;

            await _context.SaveChangesAsync();
            await transaction.CommitAsync();

            _logger.LogInformation("Created order {OrderNumber} for {CustomerEmail} (Variant Y)", orderNumber, dto.CustomerEmail);

            // Return the created order with line items
            var createdOrder = await GetByIdAsync(order.Id);
            return createdOrder!;
        }
        catch
        {
            await transaction.RollbackAsync();
            throw;
        }
    }

    /// <summary>
    /// Variant A: Dual-layer inventory with preallocated memory (99%+ zero network I/O)
    ///
    /// Architecture:
    /// - Layer 1: SPU counter (campaign-wide limit) - enforces total campaign limit
    /// - Layer 2: SKU caches (per-SKU inventory) - tracks individual variants
    ///
    /// Flow: Check SPU counter → decrement → check SKU cache → decrement → success
    /// If SPU depleted: async refill from Redis pool (non-blocking)
    ///
    /// Orders queue to Redis for write-back AFTER campaign ends.
    /// </summary>
    private async Task<OrderResponseDto> CreateOrderVariantAAsync(OrderCreateDto dto, string orderNumber, Guid flashSaleId)
    {
        _logger.LogDebug("Variant A (Dual-Layer) for order {OrderNumber}, campaign {FlashSaleId}", orderNumber, flashSaleId);

        // Reserve each SKU using dual-layer CampaignMemoryAllocator
        // SPU counter (Layer 1) ensures campaign-wide limit
        // SKU cache (Layer 2) tracks per-SKU inventory
        var reservedItems = new List<(Guid SkuId, int Quantity, decimal UnitPrice)>();
        decimal subtotal = 0;

        try
        {
            foreach (var itemDto in dto.LineItems)
            {
                decimal itemPrice = 0;
                decimal itemSubtotal = 0;

                // Reserve items using dual-layer allocator (one at a time for quantity > 1)
                for (int i = 0; i < itemDto.Quantity; i++)
                {
                    var result = _campaignAllocator.ReserveItem(flashSaleId, itemDto.SkuId);

                    if (!result.Success)
                    {
                        switch (result.PriceType)
                        {
                            case "sold_out":
                                _logger.LogWarning("SKU {SkuId} sold out in campaign {CampaignId}", itemDto.SkuId, flashSaleId);
                                throw new InvalidOperationException($"SKU {itemDto.SkuId} sold out");

                            case "not_allocated":
                                _logger.LogWarning("SKU {SkuId} not allocated to this node", itemDto.SkuId);
                                throw new InvalidOperationException($"SKU {itemDto.SkuId} not available on this server");

                            default:
                                _logger.LogError("Reservation failed for SKU {SkuId}: {PriceType}", itemDto.SkuId, result.PriceType);
                                throw new InvalidOperationException($"Failed to reserve SKU {itemDto.SkuId}");
                        }
                    }

                    // Check if fell back to ordinary stock (benchmark stop indicator)
                    if (result.PriceType == "ordinary")
                    {
                        _logger.LogWarning("[BENCHMARK STOP] SKU {SkuId} using ordinary price", itemDto.SkuId);
                    }

                    // Capture the price from first successful reservation
                    if (i == 0)
                    {
                        itemPrice = result.Price;
                    }
                    itemSubtotal += result.Price;
                }

                subtotal += itemSubtotal;
                reservedItems.Add((itemDto.SkuId, itemDto.Quantity, itemPrice));
            }

            // Queue order for async write-back (after campaign ends)
            var lineItemsPayload = reservedItems.Select(item => new Dictionary<string, object>
            {
                ["sku_id"] = item.SkuId.ToString(),
                ["quantity"] = item.Quantity,
                ["unit_price"] = item.UnitPrice.ToString("F2")
            }).ToList();

            var orderPayload = new Dictionary<string, object>
            {
                ["order_number"] = orderNumber,
                ["customer_email"] = dto.CustomerEmail,
                ["customer_name"] = dto.CustomerName ?? "",
                ["flash_sale_id"] = flashSaleId.ToString(),
                ["line_items"] = lineItemsPayload,
                ["subtotal"] = subtotal.ToString("F2"),
                ["created_at"] = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()
            };

            await _redisCache.QueueOrderAsync(orderPayload);

            _logger.LogDebug("Order {OrderNumber} reserved (Variant A dual-layer), queued for write-back", orderNumber);

            return new OrderResponseDto
            {
                OrderNumber = orderNumber,
                CustomerEmail = dto.CustomerEmail,
                CustomerName = dto.CustomerName,
                FlashSaleCampaignId = flashSaleId,
                Status = OrderStatus.Pending,
                Subtotal = subtotal,
                TotalAmount = subtotal + dto.TaxAmount + dto.ShippingAmount
            };
        }
        catch (Exception ex)
        {
            // Note: Dual-layer doesn't support easy rollback (items consumed from local cache)
            // In production, implement compensation or accept small variance
            _logger.LogError(ex, "Order {OrderNumber} failed in Variant A", orderNumber);
            throw;
        }
    }

    public async Task<OrderResponseDto?> UpdateStatusAsync(Guid id, string status)
    {
        var order = await _context.Orders.FindAsync(id);
        if (order == null) return null;

        if (!Enum.TryParse<OrderStatus>(status, true, out var orderStatus))
        {
            throw new InvalidOperationException($"Invalid order status: {status}");
        }

        order.Status = orderStatus;
        await _context.SaveChangesAsync();

        _logger.LogInformation("Updated order {OrderId} status to {Status}", id, status);
        
        return await GetByIdAsync(id);
    }

    public async Task<bool> CancelAsync(Guid id)
    {
        var order = await _context.Orders
            .Include(o => o.LineItems)
            .FirstOrDefaultAsync(o => o.Id == id);
            
        if (order == null) return false;

        // Release reserved inventory
        foreach (var lineItem in order.LineItems)
        {
            var inventory = await _context.Inventories
                .FirstOrDefaultAsync(i => i.SkuId == lineItem.SkuId);
            inventory?.ReleaseQuantity(lineItem.Quantity);
        }

        order.Status = OrderStatus.Cancelled;
        await _context.SaveChangesAsync();

        _logger.LogInformation("Cancelled order {OrderId}", id);
        
        return true;
    }

    public async Task<PaymentResponseDto> CreatePaymentAsync(Guid orderId, PaymentCreateDto dto)
    {
        var order = await _context.Orders.FindAsync(orderId);
        if (order == null)
        {
            throw new InvalidOperationException("Order not found");
        }

        var payment = new Payment
        {
            OrderId = orderId,
            Amount = dto.Amount,
            Currency = dto.Currency,
            PaymentMethod = dto.PaymentMethod,
            ReferenceNumber = dto.ReferenceNumber,
            Notes = dto.Notes
        };

        // Simulate payment processing
        if (dto.Amount > 0)
        {
            payment.Status = PaymentStatus.Captured;
            payment.GatewayTransactionId = $"txn_{DateTimeOffset.UtcNow.ToUnixTimeSeconds()}";
            
            // Update order status
            order.Status = OrderStatus.Confirmed;
        }

        _context.Payments.Add(payment);
        await _context.SaveChangesAsync();

        _logger.LogInformation("Created payment {PaymentId} for order {OrderId}", payment.Id, orderId);

        return _mapper.Map<PaymentResponseDto>(payment);
    }
}

File: variant-a/java-service/src/main/java/com/flashsale/api/service/OrderService.java
package com.flashsale.api.service;

import com.flashsale.api.dto.OrderDtos.*;
import com.flashsale.api.entity.*;
import com.flashsale.api.repository.*;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDateTime;
import java.util.*;
import java.util.AbstractMap;
import java.util.stream.Collectors;

@Service
@Transactional
public class OrderService {

    private final OrderRepository orderRepository;
    private final OrderLineItemRepository orderLineItemRepository;
    private final PaymentRepository paymentRepository;
    private final SkuRepository skuRepository;
    private final InventoryRepository inventoryRepository;
    private final FlashSaleRepository flashSaleRepository;
    private final SnowflakeIdGenerator idGenerator;
    private final RedisCacheService redisCache;
    private final AdaptiveInventoryManager adaptiveInventory;
    private final AllocationManagerV2 allocationManagerV2;
    private final CampaignMemoryAllocator campaignAllocator;

    @Autowired
    public OrderService(
            OrderRepository orderRepository,
            OrderLineItemRepository orderLineItemRepository,
            PaymentRepository paymentRepository,
            SkuRepository skuRepository,
            InventoryRepository inventoryRepository,
            FlashSaleRepository flashSaleRepository,
            SnowflakeIdGenerator idGenerator,
            RedisCacheService redisCache,
            AdaptiveInventoryManager adaptiveInventory,
            AllocationManagerV2 allocationManagerV2,
            CampaignMemoryAllocator campaignAllocator) {
        this.orderRepository = orderRepository;
        this.orderLineItemRepository = orderLineItemRepository;
        this.paymentRepository = paymentRepository;
        this.skuRepository = skuRepository;
        this.inventoryRepository = inventoryRepository;
        this.flashSaleRepository = flashSaleRepository;
        this.idGenerator = idGenerator;
        this.redisCache = redisCache;
        this.adaptiveInventory = adaptiveInventory;
        this.allocationManagerV2 = allocationManagerV2;
        this.campaignAllocator = campaignAllocator;
    }

    @Transactional(readOnly = true)
    public Page<OrderResponseDto> getAllOrders(int page, int size, String customerEmail) {
        Pageable pageable = PageRequest.of(page, size, Sort.by("createdAt").descending());
        Page<Order> orders = orderRepository.findAllWithFilter(customerEmail, pageable);
        return orders.map(this::mapToResponseDto);
    }

    @Transactional(readOnly = true)
    public Optional<OrderResponseDto> getOrderById(UUID id) {
        return orderRepository.findByIdWithLineItems(id)
                .map(this::mapToResponseDto);
    }

    /**
     * Create order with intelligent routing:
     * - If SKU is in active flash sale campaign → Use Variant A (In-memory allocation)
     * - Otherwise → Use Variant Y (database transaction)
     * Frontend sees same API, backend handles routing transparently.
     *
     * OPTIMIZED: Uses CampaignMemoryAllocator.getCampaignForSku() instead of Redis lookup.
     * This eliminates 1 Redis round-trip per request on the hot path.
     */
    public OrderResponseDto createOrder(OrderCreateDto createDto) {
        String orderNumber = String.format("ORD-%d", idGenerator.generate());

        // Step 1: Check if ANY SKU is in active flash sale campaign
        // FAST PATH: Check allocator directly (zero Redis I/O)
        List<UUID> skuIds = createDto.getLineItems().stream()
                .map(OrderLineItemCreateDto::getSkuId)
                .collect(Collectors.toList());

        UUID flashSaleId = null;
        boolean useVariantA = false;

        for (UUID skuId : skuIds) {
            // Zero Redis I/O - checks in-memory campaign registry
            CampaignMemoryAllocator.CampaignLookupResult result = campaignAllocator.getCampaignForSku(skuId);
            if (result.isLoaded) {
                flashSaleId = result.campaignId;
                useVariantA = true;
                break;
            }
        }

        if (useVariantA) {
            // VARIANT A: In-Memory Allocation (Flash Sale Path)
            return createOrderVariantA(createDto, orderNumber, flashSaleId);
        } else {
            // VARIANT Y: Database Transaction (Regular Order Path)
            return createOrderVariantY(createDto, orderNumber);
        }
    }

    /**
     * Variant Y: Traditional database transaction path (4-7 queries per order).
     * Used for regular orders when SKU is NOT in active flash sale.
     */
    private OrderResponseDto createOrderVariantY(OrderCreateDto createDto, String orderNumber) {
        // Create order
        Order order = new Order();
        order.setOrderNumber(orderNumber);
        order.setCustomerEmail(createDto.getCustomerEmail());
        order.setCustomerName(createDto.getCustomerName());
        order.setTaxAmount(createDto.getTaxAmount());
        order.setShippingAmount(createDto.getShippingAmount());
        order.setCurrency(createDto.getCurrency());
        order.setNotes(createDto.getNotes());
        order.setFlashSaleCampaignId(createDto.getFlashSaleCampaignId());

        order = orderRepository.save(order); // Get the order ID

        // Add line items and calculate totals
        BigDecimal subtotal = BigDecimal.ZERO;
        for (OrderLineItemCreateDto itemDto : createDto.getLineItems()) {
            // Get SKU information
            Sku sku = skuRepository.findByIdWithInventoryAndSpu(itemDto.getSkuId())
                    .orElseThrow(() -> new IllegalArgumentException("SKU " + itemDto.getSkuId() + " not found"));

            // Check inventory
            if (sku.getTrackInventory() && sku.getInventory() != null) {
                if (!sku.getInventory().canFulfillQuantity(itemDto.getQuantity())) {
                    throw new IllegalStateException("Insufficient inventory for SKU " + sku.getSkuCode());
                }

                // Reserve inventory
                sku.getInventory().reserveQuantity(itemDto.getQuantity());
                inventoryRepository.save(sku.getInventory());
            }

            // Check for active flash sale campaign and apply flash price
            FlashSale activeCampaign = null;
            if (sku.getSpuId() != null) {
                LocalDateTime now = LocalDateTime.now();
                List<FlashSale> campaigns = flashSaleRepository.findAll();
                activeCampaign = campaigns.stream()
                    .filter(c -> c.getSpuId().equals(sku.getSpuId()))
                    .filter(c -> c.getIsActive())
                    .filter(c -> "active".equals(c.getStatus()))
                    .filter(c -> !c.getStartTime().isAfter(now) && !c.getEndTime().isBefore(now))
                    .findFirst()
                    .orElse(null);
            }

            // Create line item - use flash price if campaign is active
            BigDecimal unitPrice;
            if (activeCampaign != null) {
                unitPrice = itemDto.getUnitPrice() != null ? itemDto.getUnitPrice() : activeCampaign.getFlashPrice();
            } else {
                unitPrice = itemDto.getUnitPrice() != null ? itemDto.getUnitPrice() : sku.getPrice();
            }
            BigDecimal totalPrice = unitPrice.multiply(BigDecimal.valueOf(itemDto.getQuantity()));

            OrderLineItem lineItem = new OrderLineItem();
            lineItem.setOrderId(order.getId());
            lineItem.setSkuId(sku.getId());
            lineItem.setQuantity(itemDto.getQuantity());
            lineItem.setUnitPrice(unitPrice);
            lineItem.setTotalPrice(totalPrice);
            lineItem.setProductName(sku.getSpu() != null ? sku.getSpu().getName() : (sku.getName() != null ? sku.getName() : "Product"));
            lineItem.setSkuCode(sku.getSkuCode());

            orderLineItemRepository.save(lineItem);
            subtotal = subtotal.add(totalPrice);
        }

        // Update order totals
        order.setSubtotal(subtotal);
        order.setTotalAmount(subtotal.add(order.getTaxAmount()).add(order.getShippingAmount()));
        order = orderRepository.save(order);

        return getOrderById(order.getId()).orElseThrow();
    }

    /**
     * Variant A: Dual-layer inventory with preallocated memory (99%+ zero network I/O)
     *
     * Architecture:
     * - Layer 1: SPU counter (campaign-wide limit) - enforces total campaign limit
     * - Layer 2: SKU caches (per-SKU inventory) - tracks individual variants
     *
     * Flow: Check SPU counter → decrement → check SKU cache → decrement → success
     * If SPU depleted: async refill from Redis pool (non-blocking)
     *
     * Orders queue to local buffer for fire-and-forget write-back.
     *
     * CRITICAL: NOT_SUPPORTED propagation prevents DB transaction overhead for in-memory operations
     */
    @Transactional(propagation = Propagation.NOT_SUPPORTED)
    private OrderResponseDto createOrderVariantA(OrderCreateDto createDto, String orderNumber, UUID flashSaleId) {
        // Track reservation results
        List<ReservedItem> reservedItems = new ArrayList<>();
        BigDecimal totalAmount = BigDecimal.ZERO;

        try {
            // Reserve each line item using CampaignMemoryAllocator
            for (OrderLineItemCreateDto itemDto : createDto.getLineItems()) {
                UUID skuId = itemDto.getSkuId();
                int quantity = itemDto.getQuantity();
                BigDecimal itemSubtotal = BigDecimal.ZERO;

                // Reserve items one by one using dual-layer CampaignMemoryAllocator
                for (int i = 0; i < quantity; i++) {
                    CampaignMemoryAllocator.ReservationResult result =
                        campaignAllocator.reserveItem(flashSaleId, skuId);

                    if (!result.success) {
                        switch (result.priceType) {
                            case "sold_out":
                                throw new IllegalStateException("SKU " + skuId + " sold out");
                            case "not_allocated":
                                throw new IllegalStateException("SKU " + skuId + " not available on this server");
                            case "not_loaded":
                                throw new IllegalStateException("Campaign " + flashSaleId + " not loaded");
                            case "ordinary":
                                throw new IllegalStateException("Campaign " + flashSaleId + " sold out");
                            default:
                                throw new IllegalStateException("Failed to reserve SKU " + skuId);
                        }
                    }

                    itemSubtotal = itemSubtotal.add(result.price);
                }

                totalAmount = totalAmount.add(itemSubtotal);
                reservedItems.add(new ReservedItem(
                    skuId, quantity, "flash",
                    itemSubtotal.divide(BigDecimal.valueOf(quantity), 2, java.math.RoundingMode.HALF_UP)
                ));
            }

            // FIRE-AND-FORGET: Queue to local buffer (zero blocking)
            Map<String, Object> orderPayload = new HashMap<>();
            orderPayload.put("order_number", orderNumber);
            orderPayload.put("customer_email", createDto.getCustomerEmail());
            orderPayload.put("customer_name", createDto.getCustomerName() != null ? createDto.getCustomerName() : "");
            orderPayload.put("flash_sale_id", flashSaleId.toString());
            orderPayload.put("total_amount", totalAmount.toString());
            orderPayload.put("created_at", System.currentTimeMillis());

            List<Map<String, Object>> lineItems = new ArrayList<>();
            for (ReservedItem item : reservedItems) {
                Map<String, Object> lineItem = new HashMap<>();
                lineItem.put("sku_id", item.skuId.toString());
                lineItem.put("quantity", item.quantity);
                lineItem.put("unit_price", item.unitPrice.toString());
                lineItems.add(lineItem);
            }
            orderPayload.put("line_items", lineItems);

            campaignAllocator.queueOrderFireAndForget(orderPayload);

            // Build immediate response
            OrderResponseDto response = new OrderResponseDto();
            response.setId(UUID.randomUUID());
            response.setOrderNumber(orderNumber);
            response.setCustomerEmail(createDto.getCustomerEmail());
            response.setCustomerName(createDto.getCustomerName());
            response.setSubtotal(totalAmount);
            response.setTaxAmount(BigDecimal.ZERO);
            response.setShippingAmount(BigDecimal.ZERO);
            response.setTotalAmount(totalAmount);
            response.setCurrency(createDto.getCurrency() != null ? createDto.getCurrency() : "USD");
            response.setStatus(OrderStatus.pending);
            response.setFlashSaleCampaignId(flashSaleId);
            response.setCreatedAt(LocalDateTime.now());
            response.setUpdatedAt(LocalDateTime.now());

            return response;

        } catch (Exception e) {
            throw e;
        }
    }


    /**
     * Helper class to track reserved items
     */
    private static class ReservedItem {
        final UUID skuId;
        final int quantity;
        final String priceType;
        final BigDecimal unitPrice;

        ReservedItem(UUID skuId, int quantity, String priceType, BigDecimal unitPrice) {
            this.skuId = skuId;
            this.quantity = quantity;
            this.priceType = priceType;
            this.unitPrice = unitPrice;
        }
    }

    public Optional<OrderResponseDto> updateOrderStatus(UUID id, String status) {
        try {
            OrderStatus orderStatus = OrderStatus.valueOf(status.toUpperCase());
            
            return orderRepository.findById(id)
                    .map(order -> {
                        order.setStatus(orderStatus);
                        orderRepository.save(order);
                        return mapToResponseDto(order);
                    });
        } catch (IllegalArgumentException e) {
            throw new IllegalArgumentException("Invalid order status: " + status);
        }
    }

    public boolean cancelOrder(UUID id) {
        Optional<Order> orderOpt = orderRepository.findByIdWithLineItems(id);
        if (orderOpt.isEmpty()) {
            return false;
        }

        Order order = orderOpt.get();
        
        // Release reserved inventory
        for (OrderLineItem lineItem : order.getLineItems()) {
            inventoryRepository.findBySkuId(lineItem.getSkuId())
                    .ifPresent(inventory -> {
                        inventory.releaseQuantity(lineItem.getQuantity());
                        inventoryRepository.save(inventory);
                    });
        }

        order.setStatus(OrderStatus.cancelled);
        orderRepository.save(order);
        
        return true;
    }

    public PaymentResponseDto createPayment(UUID orderId, PaymentCreateDto dto) {
        Order order = orderRepository.findById(orderId)
                .orElseThrow(() -> new IllegalArgumentException("Order not found"));

        Payment payment = new Payment();
        payment.setOrderId(orderId);
        payment.setAmount(dto.getAmount());
        payment.setCurrency(dto.getCurrency());
        payment.setPaymentMethod(dto.getPaymentMethod());
        payment.setReferenceNumber(dto.getReferenceNumber());
        payment.setNotes(dto.getNotes());

        // Simulate payment processing
        if (dto.getAmount().compareTo(BigDecimal.ZERO) > 0) {
            payment.setStatus(PaymentStatus.captured);
            payment.setGatewayTransactionId("txn_" + Instant.now().getEpochSecond());
            
            // Update order status
            order.setStatus(OrderStatus.confirmed);
            orderRepository.save(order);
        }

        payment = paymentRepository.save(payment);
        return mapPaymentToResponseDto(payment);
    }

    private OrderResponseDto mapToResponseDto(Order order) {
        OrderResponseDto dto = new OrderResponseDto();
        dto.setId(order.getId());
        dto.setOrderNumber(order.getOrderNumber());
        dto.setCustomerEmail(order.getCustomerEmail());
        dto.setCustomerName(order.getCustomerName());
        dto.setSubtotal(order.getSubtotal());
        dto.setTaxAmount(order.getTaxAmount());
        dto.setShippingAmount(order.getShippingAmount());
        dto.setTotalAmount(order.getTotalAmount());
        dto.setCurrency(order.getCurrency());
        dto.setStatus(order.getStatus());
        dto.setNotes(order.getNotes());
        dto.setFlashSaleCampaignId(order.getFlashSaleCampaignId());
        dto.setCreatedAt(order.getCreatedAt());
        dto.setUpdatedAt(order.getUpdatedAt());

        List<OrderLineItemResponseDto> lineItemDtos = order.getLineItems().stream()
                .map(this::mapLineItemToResponseDto)
                .collect(Collectors.toList());
        dto.setLineItems(lineItemDtos);

        return dto;
    }

    private OrderLineItemResponseDto mapLineItemToResponseDto(OrderLineItem lineItem) {
        OrderLineItemResponseDto dto = new OrderLineItemResponseDto();
        dto.setId(lineItem.getId());
        dto.setSkuId(lineItem.getSkuId());
        dto.setQuantity(lineItem.getQuantity());
        dto.setUnitPrice(lineItem.getUnitPrice());
        dto.setTotalPrice(lineItem.getTotalPrice());
        dto.setProductName(lineItem.getProductName());
        dto.setSkuCode(lineItem.getSkuCode());
        dto.setCreatedAt(lineItem.getCreatedAt());
        return dto;
    }

    private PaymentResponseDto mapPaymentToResponseDto(Payment payment) {
        PaymentResponseDto dto = new PaymentResponseDto();
        dto.setId(payment.getId());
        dto.setOrderId(payment.getOrderId());
        dto.setAmount(payment.getAmount());
        dto.setCurrency(payment.getCurrency());
        dto.setPaymentMethod(payment.getPaymentMethod());
        dto.setGatewayTransactionId(payment.getGatewayTransactionId());
        dto.setStatus(payment.getStatus());
        dto.setReferenceNumber(payment.getReferenceNumber());
        dto.setNotes(payment.getNotes());
        dto.setCreatedAt(payment.getCreatedAt());
        dto.setUpdatedAt(payment.getUpdatedAt());
        return dto;
    }
}
